aws/src/main/java/org/apache/iceberg/aws/glue/DynamoLockManager.java

```diff
@@ -0,0 +1,373 @@
+      }
+    }
+
+    return succeeded;
+  }
+
+  private static Map<String, AttributeValue> toKey(String entityId) {
+    return ImmutableMap.of(COL_LOCK_ENTITY_ID, AttributeValue.builder().s(entityId).build());
+  }
+
+  private static Map<String, AttributeValue> toNewItem(String entityId, String ownerId, long heartbeatTimeoutMs) {
+    return ImmutableMap.of(
+        COL_LOCK_ENTITY_ID, AttributeValue.builder().s(entityId).build(),
+        COL_LOCK_OWNER_ID, AttributeValue.builder().s(ownerId).build(),
+        COL_VERSION, AttributeValue.builder().s(UUID.randomUUID().toString()).build(),
+        COL_LEASE_DURATION_MS, AttributeValue.builder().n(Long.toString(heartbeatTimeoutMs)).build());
+  }
+
+  private static Map<String, AttributeValue> toLockIdValues(String entityId, String ownerId) {
+    return ImmutableMap.of(
+        ":eid", AttributeValue.builder().s(entityId).build(),
+        ":oid", AttributeValue.builder().s(ownerId).build());
+  }
+
+  @Override
+  public void close() {
+    dynamo.close();
+    heartbeats.values().forEach(DynamoHeartbeat::cancel);
+    heartbeats.clear();
+  }
+
+  /**
+   * The lock table schema, for users who would like to create the table separately
+   * @return lock table schema
+   */
+  public static List<KeySchemaElement> lockTableSchema() {
+    return LOCK_TABLE_SCHEMA;
+  }
+
+  /**
+   * The lock table column definition, for users who whould like to create the table separately
+   * @return lock table column definition
+   */
+  public static List<AttributeDefinition> lockTableColDefinitions() {
+    return LOCK_TABLE_COL_DEFINITIONS;
+  }
+
+  private static class DynamoHeartbeat implements Runnable {
+
+    private final DynamoDbClient dynamo;
+    private final String lockTableName;
+    private final long intervalMs;
+    private final long timeoutMs;
+    private final String entityId;
+    private final String ownerId;
+    private ScheduledFuture<?> future;
+
+    DynamoHeartbeat(DynamoDbClient dynamo, String lockTableName, long intervalMs, long timeoutMs,
+                           String entityId, String ownerId) {
+      this.dynamo = dynamo;
+      this.lockTableName = lockTableName;
+      this.intervalMs = intervalMs;
+      this.timeoutMs = timeoutMs;
+      this.entityId = entityId;
+      this.ownerId = ownerId;
+      this.future = null;
+    }
+
+    @Override
+    public void run() {
+      try {
+        dynamo.putItem(PutItemRequest.builder()
+            .tableName(lockTableName)
+            .item(toNewItem(entityId, ownerId, timeoutMs))
+            .conditionExpression(CONDITION_LOCK_ID_MATCH)
+            .expressionAttributeValues(toLockIdValues(entityId, ownerId))
+            .build());
+      } catch (ConditionalCheckFailedException e) {
+        LOG.error("Fail to heartbeat for entity: {}, owner: {} due to conditional check failure, " +
+                "unsafe concurrent commits might be going on", entityId, ownerId, e);
+      } catch (DynamoDbException e) {
```

Comment:
This is one of the few times I think that we actually need to broaden the catch clause to `Exception` or `RuntimeException` instead of `DynamoDbException`. If anything happens that causes the heartbeat to fail, like an unrelated `NullPointerException` inside the AWS SDK, then we want this to keep trying as long as the process thinks it holds the lock. Catching `RuntimeException` ensures that no exceptions should stop this from getting scheduled again.